lib/Re2.fx

```diff
@@ -0,0 +1,828 @@
+            fold filtered = [] for piece <- pieces_list
+            {   
+                val ok = match piece { | RPString simple_piece => String.length(simple_piece)!=0  | RPInt sub_num => true}
+                if (ok) {piece::filtered} else {filtered}
+            }                
+        }
+        else
+        {
+            RPString(rewrite) :: []
+        }
+    
+    val fold max_sub = -1 for piece <- piece_lst
+    {   
+        match piece
+        {
+            | RPString simple_piece => max_sub
+            | RPInt sub_num => max(max_sub, sub_num) 
+        }
+    }    
+    replace_pattern_t {pieces = piece_lst, max_subnum = max_sub}
+}
+
+//Service function
+fun compose_replacement(cloth: string, french_curve: replace_pattern_t, found_subs: (int, int)[]) : string 
+{
+    val fold rlist = [] for piece <- french_curve.pieces
+        {   
+            match piece
+            {
+                | RPString simple_piece => simple_piece::rlist
+                | RPInt sub_num => 
+                    val (sub_start, sub_end) = found_subs[sub_num]
+                    cloth[sub_start:sub_end]::rlist
+            }
+        }
+    join("", List.rev(rlist))
+}
+
+
+//                                                                 (success, result)
+fun replace(str: string, re: regex_t, rewrite: replace_pattern_t): (bool,    string)
+{
+    if number_of_capturing_groups(regex_t {handle = re.find_r, find_r = re.find_r}) <= rewrite.max_subnum
+        {throw BadArgError} //[TODO]: BadArg or something more detailed? 
+    val (is_found, ranges) = find(str, re)
+    (is_found,
+        if(is_found)
+        {
+            val (fullmatch_start, fullmatch_end) = ranges[0]
+            val j_list = str[:fullmatch_start] :: compose_replacement(str, rewrite, ranges) ::str[fullmatch_end:] :: []
+            String.join("", j_list)
+        }
+        else
+        {str})
+}
+
+//                                                                        (success, result)
+fun global_replace(str: string, re: regex_t, rewrite: replace_pattern_t): (bool,    string)
+{
+    if number_of_capturing_groups(regex_t {handle = re.find_r, find_r = re.find_r}) <= rewrite.max_subnum
+        {throw BadArgError} //[TODO]: BadArg or something more detailed? 
+
+    val (is_found, ranges) = findall(str, re)
+    (is_found,
+        if(is_found)
+        {
+            val (len_i, len_j) = size(ranges) //[TODO]: Attempt to call size on matrix causes C compile error.
+            //println(len_i)
+            //""
+            var str_list = []
+            var pos = 0
+            for i <- 0: len_i
+            {   
+                val (substart,subend) = ranges[i,0]
+                str_list = compose_replacement(str, rewrite, ranges[i,:]) :: str[pos:substart] :: str_list
+                pos = subend
+            }
+            str_list = str[pos:] :: str_list
+            String.join("", List.rev(str_list))
```

Comment:
it should be possible now to use ```"".join(str_list.rev())``` notation

Reply:
I left this unchanged. Looks clearer.